pkg/util/modules/modules.go

```diff
@@ -23,19 +26,33 @@ type Manager struct {
 	modules map[string]*module
 }
 
+// ModuleOption contains options of a module.
+// public option indicates if a module can be use as top level module (i.e. pass in as command line argument).
+type ModuleOption struct {
+	Public bool
+}
+
 // NewManager creates a new Manager
 func NewManager() *Manager {
 	return &Manager{
 		modules: make(map[string]*module),
 	}
 }
 
-// RegisterModule registers a new module with name and init function
+// RegisterModule registers a new module with name and init function. By default a module is public.
 // name must be unique to avoid overwriting modules
 // if initFn is nil, the module will not initialise
 func (m *Manager) RegisterModule(name string, initFn func() (services.Service, error)) {
+	m.RegisterModuleWithOption(name, initFn, defaultModuleOption())
+}
+
+// RegisterModule registers a new module with name, init function, and option
+// name must be unique to avoid overwriting modules
+// if initFn is nil, the module will not initialise
+func (m *Manager) RegisterModuleWithOption(name string, initFn func() (services.Service, error), option ModuleOption) {
```

Comment:
I would prefer to have single `RegisterModule` method, with variable arguments with individual options.

We can use [functional options](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis) pattern for this.

Reply:
I would prefer single `RegisterModule` too, I created `RegisterModuleWithOptions` just because Go doesn't allow method overloading (I came from Java :)) and I did not want to break backward compatibility. Thanks for pointing me to Functional options, I learned something new. 

But [this StackOverflow post](https://stackoverflow.com/questions/55163005/could-adding-variadic-parameters-to-a-function-break-existing-code) gave some examples where adding variadic parameter may break existing code; should we worry about this?

Reply:
Issues pointed out by StackOverflow post are real, but not something to be worried about in this case.